core/src/main/java/org/opensearch/remote/metadata/client/AbstractSdkClient.java

```diff
@@ -66,4 +99,129 @@ protected boolean shouldUseId(String id) {
+        });
+    }
+
+    /**
+     * Add the global resource to cache.
+     * @param request origin request.
+     * @param dataFetchedWithGlobalTenantId data fetched with global tenant id, which means this resource is a global resource.
+     * @return {@link CompletionStage<GetDataObjectResponse>} The response been added to cache and then replaced the tenant_id to request's tenant id.
+     */
+    protected CompletionStage<GetDataObjectResponse> addToGlobalResourceCache(
+        GetDataObjectRequest request,
+        CompletionStage<GetDataObjectResponse> dataFetchedWithGlobalTenantId
+    ) {
+        return dataFetchedWithGlobalTenantId.thenCompose(res -> {
+            if (globalResourceCacheTTL.getMillis() == 0) {
+                return CompletableFuture.completedFuture(replaceGlobalTenantId(request, res));
+            }
+            GLOBAL_RESOURCES_CACHE.put(buildGlobalCacheKey(request.index(), request.id()), new Tuple<>(res, System.currentTimeMillis()));
+            return CompletableFuture.completedFuture(getGlobalResourceDataFromCache(request));
+        });
+    }
+
+    /**
+     * Read global resource from cache.
+     * @param request origin get request.
+     * @return {@link GetDataObjectResponse} The response cached in cache and replaced with request tenant id.
+     */
+    protected GetDataObjectResponse getGlobalResourceDataFromCache(GetDataObjectRequest request) {
+        String checkingKey = buildGlobalCacheKey(request.index(), request.id());
+        if (!GLOBAL_RESOURCES_CACHE.containsKey(checkingKey)) {
+            return null;
+        }
+        Tuple<GetDataObjectResponse, Long> tuple = GLOBAL_RESOURCES_CACHE.get(checkingKey);
+        // Replace the tenant id in the global resource response to actual tenant id to bypass the validation of the resources:
+        // e.g.:
+        // https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLAgentExecutor.java#L206
+        if (System.currentTimeMillis() - tuple.v2() > globalResourceCacheTTL.getMillis()) {
+            // Cache is expired case.
+            GLOBAL_RESOURCES_CACHE.remove(checkingKey);
+            return null;
+        }
+        return replaceGlobalTenantId(request, tuple.v1());
+    }
+
+    private GetDataObjectResponse replaceGlobalTenantId(GetDataObjectRequest request, GetDataObjectResponse response) {
+        response.source().put(TENANT_ID_FIELD_KEY, request.tenantId());
+        GetResponse getResponse = response.getResponse();
+        if (getResponse == null) {
+            throw new OpenSearchStatusException(
+                "Cached response is null, please check configuration with system admin!",
+                RestStatus.INTERNAL_SERVER_ERROR
+            );
+        }
+        String responseStr = getResponse.toString()
+            .replaceAll(globalTenantId, Optional.ofNullable(request.tenantId()).orElse(DEFAULT_TENANT));
```

Comment:
Not really fond of blindly replacing strings that could in theory match places other than the tenant ID field.

It looks like we're controlling the tenant ID in the source (line 169) so where else are we expecting it?  And couldn't we just have better control of the builder after parsing but before we return it below on line 185?

This also reminds me that I need to check whether we validate that the global tenant ID doesn't break XContent parsing.

Reply:
> To keep the global tenant ids are all replaced with user tenant id, we have to manually create a parser and a source and set them into a new GetDataObjectResponse.

My question is, why can't we create a `GetDataObjectResponse` and then just replace its TenantID before returning it?  Pretty sure the builder can be initiated with a previous response and any individual field swapped out?